source/language/subroutines.rst

```diff
@@ -70,4 +70,65 @@ this
    c = measure q;
    c2 = measure r;
    bit result;
-   result = parity(c ++ c2);
+   result = parity(c || c2);
+
+.. _arrays-in-subroutines:
+
+Arrays in subroutines
+---------------------
+
+Arrays may be passed as parameters to subroutines and externs. All array
+parameters are passed as references and must include a type modifier specifying
+if the parameter is ``const`` or ``mutable``. The number of dimensions for all
+array parameters must be specified using the ``#dim = literal/const identifier``
+syntax below, or specific lengths for each dimension must be provided.
+The unspecified-length version is provided because the lengths of
+the dimensions of array parameters (in the case of strided access) may not be
+known until runtime. Passing multiple overlapping mutable references to the same
+array to a subroutine is forbidden. However, the compiler will not always be
+able to resolve when this happens, and if it does, then no guarantees are made
+about the order that updates are made in.
+
+.. code-block:: c
+
+   def specified_sub(const array[int[8], 2, 10] arr_arg) {...}
+   def arr_subroutine(const array[int[8], #dim = 1] arr_arg) {...}
+   def mut_subroutine(mutable array[int[8], #dim = 1] arr_arg) {
+     arr_arg[2] = 10; // allowed
+     ...
+   }
+   array[int[8], 5] aa;
+   array[int[8], 3, 5] bb;
+
+   arr_subroutine(aa);
+   arr_subroutine(bb[1, 0:3]);
+   mut_subroutine(aa[1:3]); // aa[3] = 10 
+
+The lifetime of the array reference is limited to within the scope of the
+subroutine definition, but it should be noted that since arrays are not
+dynamically allocated the memory associated with the array stays intact after
+subroutine exit. Additionally, the OpenQASM3 language is not anticipated to
+support explicit user-controlled creation of pointers and references outside
+of the specific context of passing arrays to subroutines.
+
+The dimensions of arrays may be queried inside of subroutines using the built-in
+``sizeof()`` function, which takes two parameters: the array being queried, and
+the zero-based dimension number requested. If the second parameter is omitted,
+then it defaults to ``0``, *i.e.* ``sizeof(arr) == sizeof(arr, 0)``.
+``sizeof()`` returns a ``const int[32]`` representing the length of the
```

Comment:
Wouldn't returning an unsigned type from `sizeof()` be more natural? i.e. `const uint[32]` or even `const uint`?